File: Sources/Worklings/GitCommitWatcher.swift
import CompanionCore
import Foundation

/// Watches the `.git` directory of each connected repository and turns forward
/// commits into `milestone` events for the pet.
///
/// Mirrors `ActivityInboxMonitor`'s discipline: FSEvents fire on the main
/// queue, the (brief) git work runs off the main actor, and only commit
/// identifiers and ancestry are ever read — never a message or diff (see
/// `GitCommitDelta`). Connecting a repository is itself the opt-in, so this is
/// independent of the "Accept Work Tool Events" inbox toggle.
///
/// No retro-credit: on connect and on start, the current HEAD is recorded as a
/// silent baseline, so only commits made while the app is watching count. A
/// `.git` directory is chatty (index and lock churn), so each repo's check is
/// debounced and coalesced.
@MainActor
final class GitCommitWatcher {
    private let session: PetSession
    private let registry: ConnectedRepoRegistry
    private var sources: [String: DispatchSourceFileSystemObject] = [:]
    private var pendingChecks: [String: Task<Void, Never>] = [:]
    /// When each repo's baseline was last established, used as the `--since`
    /// boundary so only commits made after that moment (i.e. while watching)
    /// are counted. In-memory: at launch the baseline resyncs to the current
    /// HEAD and this resets to now, so a relaunch never retro-counts.
    private var lastCheckedAt: [String: Date] = [:]
    private var isRunning = false

    private static let debounce: Duration = .milliseconds(600)

    init(session: PetSession, registry: ConnectedRepoRegistry = ConnectedRepoRegistry()) {
        self.session = session
        self.registry = registry
    }

    func connectedRepoPaths() -> [String] {
        registry.all().map(\.path)
    }

    /// Begins watching every connected repo, resyncing each baseline to its
    /// current HEAD without emitting — commits made while the app was closed
    /// are not retro-credited.
    func start() {
        guard !isRunning else {
            return
        }
        isRunning = true
        for repo in registry.all() {
            resyncBaselineAndWatch(path: repo.path)
        }
    }

    func stop() {
        isRunning = false
        for source in sources.values {
            source.cancel()
        }
        sources.removeAll()
        for task in pendingChecks.values {
            task.cancel()
        }
        pendingChecks.removeAll()
    }

    /// Connects a repo: records a silent baseline at the current HEAD and starts
    /// watching. Returns false (and does nothing) if the path is not a git
    /// repository. A no-op if already connected.
    @discardableResult
    func connect(path: String) async -> Bool {
        // Canonicalize to the repo top-level (off the main actor, so the
        // connecting click is never blocked on git), so connecting a
        // subdirectory, a symlink, or a `..`-laden path can't register the same
        // repository twice and double its rewards.
        guard let root = await Self.resolveTopLevel(path) else {
            return false
        }
        guard !registry.contains(path: root) else {
            return true
        }
        // Baseline starts nil and is set to the current HEAD by the off-main
        // resync below, so we never retro-credit history already behind the repo.
        registry.add(path: root, lastSeenSHA: nil)
        if isRunning {
            resyncBaselineAndWatch(path: root)
        }
        return true
    }

    private nonisolated static func resolveTopLevel(_ path: String) async -> String? {
        GitRepository.topLevel(atPath: path)
    }

    func disconnect(path: String) {
        sources[path]?.cancel()
        sources[path] = nil
        pendingChecks[path]?.cancel()
        pendingChecks[path] = nil
        lastCheckedAt[path] = nil
        registry.remove(path: path)
    }

    /// Resolves HEAD and the `.git` directory off the main actor, then applies
    /// the baseline and installs the watch back on main. No git call ever runs
    /// on the main thread here, so neither launch nor a connect click can be
    /// blocked by a slow or wedged repository.
    private func resyncBaselineAndWatch(path: String) {
        Task { [weak self] in
            let info = await Self.repoInfo(path: path)
            guard let self, self.isRunning else {
                return
            }
            self.registry.updateLastSeenSHA(path: path, sha: info.head)
            self.lastCheckedAt[path] = Date()
            if let gitDirectory = info.gitDirectory {
                self.installWatch(path: path, gitDirectory: gitDirectory)
            }
        }
    }

    private nonisolated static func repoInfo(path: String) async -> (head: String?, gitDirectory: String?) {
        (GitRepository.head(atPath: path), GitRepository.gitDirectoryPath(atPath: path))
    }

    private func installWatch(path: String, gitDirectory: String) {
        guard sources[path] == nil else {
            return
        }

        let descriptor = open(gitDirectory, O_EVTONLY)
        guard descriptor >= 0 else {
            NSLog("Worklings could not open %@ for git watching.", gitDirectory)
            return
        }

        let source = DispatchSource.makeFileSystemObjectSource(
            fileDescriptor: descriptor,
            eventMask: [.write, .rename, .delete],
            queue: .main
        )
        source.setEventHandler { [weak self] in
            self?.scheduleCheck(path: path)
        }
        source.setCancelHandler {
            close(descriptor)
        }
        sources[path] = source
        source.resume()
    }

    /// Coalesces a burst of `.git` writes into one debounced check per repo.
    private func scheduleCheck(path: String) {
        let oldSHA = registry.lastSeenSHA(path: path)
        let since = lastCheckedAt[path] ?? Date.distantPast
        pendingChecks[path]?.cancel()
        pendingChecks[path] = Task { [weak self] in
            try? await Task.sleep(for: Self.debounce)
            guard !Task.isCancelled else {
                return
            }

            let outcome = await Self.evaluate(path: path, oldSHA: oldSHA, since: since)

            guard !Task.isCancelled, let self else {
                return
            }
            self.registry.updateLastSeenSHA(path: path, sha: outcome.newSHA)
            if outcome.newSHA != nil {
                self.lastCheckedAt[path] = Date()
            }
            guard outcome.milestones > 0 else {
                return
            }
            // One milestone per genuine new commit; the per-source daily count
            // in PetBrain applies the diminishing returns from here.
            let now = Date()
            for _ in 0..<outcome.milestones {
                self.session.receive(GitActivitySource.event(.milestone, at: now))
            }
        }
    }

    /// The git-touching half, run off the main actor. Pure decision lives in
    /// `GitCommitDelta`; this only gathers the facts it needs.
    private nonisolated static func evaluate(
        path: String,
        oldSHA: String?,
        since: Date
    ) async -> (newSHA: String?, milestones: Int) {
        guard let newSHA = GitRepository.head(atPath: path) else {
            return (nil, 0)
        }
        let isAncestor = oldSHA.map { GitRepository.isAncestor($0, of: newSHA, atPath: path) } ?? false
        // `from: oldSHA` may be nil — an empty-at-connect repo whose first
        // commit(s) count from the root; recentCommitCount handles that.
        let recent = GitRepository.recentCommitCount(from: oldSHA, to: newSHA, since: since, atPath: path)
        let milestones = GitCommitDelta.milestonesToEmit(
            oldSHA: oldSHA,
            newSHA: newSHA,
            oldIsAncestorOfNew: isAncestor,
            commitsAhead: recent
        )
        return (newSHA, milestones)
    }
}

File: godot/worklings/core/host/GitCommitWatcher.cs
using Godot;
using System.Collections.Generic;
using Worklings.Core.Pet;

namespace Worklings.Core.Host;

/// Watches each connected repository's HEAD and turns forward commits into
/// milestones.
///
/// **No retro-credit.** On the first pass over a repo, the current HEAD is
/// recorded as a silent baseline and nothing is emitted. Only commits made while
/// the app is actually watching ever count — which is also why the count is
/// filtered by commit date rather than by graph distance, so a `pull` that
/// fast-forwards over a hundred old commits earns exactly nothing.
///
/// The decision — what a HEAD movement represents — is `GitCommitDelta`, ported
/// and probed. This gathers the facts it needs and delivers the result.
///
/// **Polling on a background thread, not FSEvents.** The Swift app watches the
/// `.git` directory with a `DispatchSource` and debounces the churn. Godot has
/// no equivalent, so this asks every fifteen seconds — and asks from a worker
/// thread, because `git` on a network mount can hang for seconds and the pet
/// must not freeze while it does. Only the delivery happens on the main thread.
///
/// Ported in behaviour from Sources/Worklings/GitCommitWatcher.swift.
public sealed partial class GitCommitWatcher : Node
{
    /// How often to look. Fifteen seconds: a commit is not urgent, and each
    /// pass spawns a `git` process per repository.
    [Export] public double PollSeconds { get; set; } = 15;

    private readonly PetSession _session;
    private readonly ConnectedRepoRegistry _registry;

    /// When each repo's baseline was established, used as the `--since`
    /// boundary. In memory rather than on disk, so a relaunch resets it along
    /// with the baseline and can never retro-count.
    private readonly Dictionary<string, System.DateTimeOffset> _baselinedAt = new();

    /// One pass at a time. A slow repo must not stack passes behind it.
    private volatile bool _busy;

    /// What the worker found, waiting to be delivered on the main thread.
    private readonly List<Finding> _pending = new();

    private double _timer;

    private readonly record struct Finding(string Path, string? Sha, int Milestones);

    public GitCommitWatcher(PetSession session, ConnectedRepoRegistry? registry = null)
    {
        _session = session;
        _registry = registry ?? new ConnectedRepoRegistry();
    }

    public IReadOnlyList<ConnectedRepo> Connected => _registry.All();

    /// Connects a repository, or says why it could not.
    ///
    /// Canonicalised to the working tree's root first, so connecting a
    /// subdirectory or a symlinked path cannot register the same repo twice and
    /// double what it pays.
    public string? Connect(string path)
    {
        string? root = GitRepository.TopLevel(path);
        if (root is null || !GitRepository.IsRepository(root))
        {
            return $"{path} is not a git repository.";
        }
        if (_registry.Contains(root))
        {
            return $"{root} is already connected.";
        }
        _registry.Add(root);
        // Baselined on the next pass rather than here, so the one place that
        // decides "this is where we started counting" is the one place.
        GD.Print($"git: watching {root}");
        return null;
    }

    public void Disconnect(string path)
    {
        _registry.Remove(path);
        _baselinedAt.Remove(path);
    }

    public override void _Ready()
    {
        var connected = _registry.All();
        GD.Print(connected.Count == 0
            ? "git: no repositories connected."
            : $"git: {connected.Count} repositor{(connected.Count == 1 ? "y" : "ies")} connected.");
        // Immediately, so the baseline is set at launch rather than fifteen
        // seconds into it — a commit made in that window would otherwise be
        // counted against a baseline from the previous session.
        Poll();
    }

    public override void _Process(double delta)
    {
        Deliver();

        _timer -= delta;
        if (_timer > 0) return;
        _timer = PollSeconds;
        Poll();
    }

    private void Poll()
    {
        if (_busy) return;
        var repos = _registry.All();
        if (repos.Count == 0) return;

        // Snapshot everything the worker needs, so it never touches the
        // registry or the session from another thread.
        var work = new List<(string Path, string? Sha, System.DateTimeOffset? Since)>();
        foreach (var repo in repos)
        {
            // Null means "never baselined in this session", which is a distinct
            // thing from any instant — not a sentinel date. `--since=@0` makes
            // git return nothing at all rather than everything, so a sentinel
            // that could reach the command line would fail in the quiet
            // direction: no commits found, no milestones, no error.
            work.Add((repo.Path, repo.LastSeenSha,
                      _baselinedAt.TryGetValue(repo.Path, out var at)
                          ? at
                          : (System.DateTimeOffset?)null));
        }

        _busy = true;
        System.Threading.Tasks.Task.Run(() =>
        {
            var found = new List<Finding>();
            foreach (var (path, oldSha, since) in work)
            {
                found.Add(Evaluate(path, oldSha, since));
            }
            lock (_pending)
            {
                _pending.AddRange(found);
            }
            _busy = false;
        });
    }

    /// The git-touching half, off the main thread. Gathers facts only; the
    /// judgement is `GitCommitDelta`'s.
    private static Finding Evaluate(
        string path, string? oldSha, System.DateTimeOffset? since)
    {
        string? newSha = GitRepository.Head(path);
        if (newSha is null)
        {
            // An empty repo, or a git we could not reach. Either way there is
            // nothing to say and nothing to record.
            return new Finding(path, null, 0);
        }

        // No baseline in this session means the repo is being seen for the
        // first time. Record where it is and emit nothing — this is the
        // no-retro-credit rule, and it is why the first pass over a repository
        // with years of history in it is silent.
        if (since is not System.DateTimeOffset from)
        {
            return new Finding(path, newSha, 0);
        }

        bool isAncestor = oldSha is not null && GitRepository.IsAncestor(oldSha, newSha, path);
        int recent = GitRepository.RecentCommitCount(oldSha, newSha, from, path);
        int milestones = GitCommitDelta.MilestonesToEmit(
            oldSha, newSha, isAncestor, recent);
        return new Finding(path, newSha, milestones);
    }

    /// Applies whatever the worker found, on the main thread.
    private void Deliver()
    {
        List<Finding> found;
        lock (_pending)
        {
            if (_pending.Count == 0) return;
            found = new List<Finding>(_pending);
            _pending.Clear();
        }

        var now = System.DateTimeOffset.Now;
        foreach (var finding in found)
        {
            _registry.UpdateLastSeenSha(finding.Path, finding.Sha);
            if (finding.Sha is not null)
            {
                _baselinedAt[finding.Path] = now;
            }
            // One milestone per genuine new commit. The per-source daily count
            // in PetBrain applies the diminishing returns from there, so a
            // fifty-commit afternoon does not pay fifty times over.
            for (int i = 0; i < finding.Milestones; i++)
            {
                _session.Receive(GitActivitySource.Event(ActivityEventKind.Milestone, now), now);
            }
            if (finding.Milestones > 0)
            {
                GD.Print($"git: {finding.Milestones} commit"
                       + $"{(finding.Milestones == 1 ? "" : "s")} in {finding.Path}");
            }
        }
    }
}
